pgmax/fg/graph.py

```diff
@@ -45,181 +43,86 @@ class FactorGraph:
-        )
-        self._register_factor_group(factor_group, name)
+            factor_group: The FactorGroup to be added to the FactorGraph.
+            factor: The Factor to be added to the FactorGraph.
 
-    def add_factor_by_type(
-        self, variable_names: List, factor_type: type, *args, **kwargs
-    ) -> None:
-        """Function to add a single factor to the FactorGraph.
-
-        Args:
-            variable_names: A list containing the connected variable names.
-                Variable names are tuples of the type (variable_group_name, variable_name_within_variable_group)
-            factor_type: Type of factor to be added
-            args: Args to be passed to the factor_type.
-            kwargs: kwargs to be passed to the factor_type, and an optional "name" argument
-                for specifying the name of a named factor group.
-
-        Example:
-            To add an ORFactor to a FactorGraph fg, run::
-
-                fg.add_factor_by_type(
-                    variable_names=variables_names_for_OR_factor,
-                    factor_type=logical.ORFactor
-                )
-        """
-        if factor_type not in FAC_TO_VAR_UPDATES:
-            raise ValueError(
-                f"Type {factor_type} is not one of the supported factor types {FAC_TO_VAR_UPDATES.keys()}"
-            )
-
-        name = kwargs.pop("name", None)
-        variables = tuple(self._variable_group[variable_names])
-        factor = factor_type(variables, *args, **kwargs)
-        factor_group = groups.SingleFactorGroup(
-            variable_group=self._variable_group,
-            variable_names_for_factors=[variable_names],
-            factor=factor,
-        )
-        self._register_factor_group(factor_group, name)
-
-    def add_factor_group(self, factory: Callable, *args, **kwargs) -> None:
-        """Add a factor group to the factor graph
-
-        Args:
-            factory: Factory function that takes args and kwargs as input and outputs a factor group.
-            args: Args to be passed to the factory function.
-            kwargs: kwargs to be passed to the factory function, and an optional "name" argument
-                for specifying the name of a named factor group.
+        Raises:
+            ValueError: If
+                (1) Both a Factor and a FactorGroup are added
+                (2) The FactorGroup involving the same variables already exists in the FactorGraph.
         """
-        name = kwargs.pop("name", None)
-        factor_group = factory(self._variable_group, *args, **kwargs)
-        self._register_factor_group(factor_group, name)
-
-    def _register_factor_group(
-        self, factor_group: groups.FactorGroup, name: Optional[str] = None
-    ) -> None:
-        """Register a factor group to the factor graph, by updating the factor graph state.
+        if factor is None and factor_group is None:
+            raise ValueError("A Factor or a FactorGroup is required")
```

Comment:
This is pretty unintuitive. In theory we can easy handle this case (just process both). But there is also no reason to restrict to handling a single factor or a single factor group.

More generally, I think it's better if we just have a single input called factors, which can be a single factor, a single factor group, or a sequence of factor/factor groups. Inside the function we have a few if/else statements to handle various cases and add all the factors.